File: server/routes/schedule.ts
import { Router } from 'express';
import { isAuthenticated } from '../middleware/auth.js';
import { collections } from '../db/collections.js';
import { ObjectId } from 'mongodb';

const router = Router();

router.get('/health', (_req, res) => {
  res.json({ status: 'ok', timestamp: new Date().toISOString() });
});

router.get('/', isAuthenticated, async (req, res) => {
  try {
    const userId = req.user!._id;
    const { startDate, endDate } = req.query;

    const where: any = { userId };

    if (startDate || endDate) {
      where.date = {};
      if (startDate) where.date.$gte = new Date(startDate as string);
      if (endDate) where.date.$lte = new Date(endDate as string);
    }

    const schedules = await (await collections.schedules).find(where).sort({ date: -1, startTime: 1 }).toArray();
    res.json(schedules);
  } catch (error) {
    console.error('Error fetching schedules:', error);
    res.status(500).json({ error: 'Failed to fetch schedules' });
  }
});

router.post('/', isAuthenticated, async (req, res) => {
  try {
    const userId = req.user!._id;
    const { date, startTime, endTime, title, subject, notes } = req.body;

    const scheduleData = {
      userId,
      date: new Date(date),
      startTime,
      endTime,
      title: title || '',
      subject: subject || '',
      notes: notes || '',
      completed: false,
      createdAt: new Date(),
      updatedAt: new Date()
    };

    const result = await (await collections.schedules).insertOne(scheduleData);
    const schedule = await (await collections.schedules).findOne({ _id: result.insertedId });
    res.json(schedule);
  } catch (error) {
    console.error('Error creating schedule:', error);
    res.status(500).json({ error: 'Failed to create schedule' });
  }
});

router.put('/:id', isAuthenticated, async (req, res) => {
  try {
    const { id } = req.params;
    const userId = req.user!._id;
    const { date, startTime, endTime, title, subject, notes, completed } = req.body;

    const updateData: any = { updatedAt: new Date() };
    if (date) updateData.date = new Date(date);
    if (startTime) updateData.startTime = startTime;
    if (endTime) updateData.endTime = endTime;
    if (title !== undefined) updateData.title = title;
    if (subject !== undefined) updateData.subject = subject;
    if (notes !== undefined) updateData.notes = notes;
    if (completed !== undefined) updateData.completed = completed;

    const result = await (await collections.schedules).updateOne(
      { _id: new ObjectId(id), userId },
      { $set: updateData }
    );

    if (result.matchedCount === 0) {
      return res.status(404).json({ error: 'Schedule not found' });
    }

    res.json({ success: true });
  } catch (error) {
    console.error('Error updating schedule:', error);
    res.status(500).json({ error: 'Failed to update schedule' });
  }
});

router.delete('/:id', isAuthenticated, async (req, res) => {
  try {
    const { id } = req.params;
    const userId = req.user!._id;

    const result = await (await collections.schedules).deleteOne({ _id: new ObjectId(id), userId });

    if (result.deletedCount === 0) {
      return res.status(404).json({ error: 'Schedule not found' });
    }

    res.json({ success: true });
  } catch (error) {
    console.error('Error deleting schedule:', error);
    res.status(500).json({ error: 'Failed to delete schedule' });
  }
});

export default router;

File: server/src/routes/schedule.rs
use crate::{
    middleware::auth::{is_authenticated, AuthUser},
    models::Schedule,
    AppState,
};
use axum::{
    extract::{Path, State},
    http::StatusCode,
    routing::{get, patch},
    Extension, Json, Router,
};
use chrono::{DateTime, Utc};
use futures::stream::TryStreamExt;
use mongodb::bson::{doc, oid::ObjectId};
use serde::Deserialize;
use std::{str::FromStr, sync::Arc};

pub fn router(state: Arc<AppState>) -> Router<Arc<AppState>> {
    Router::new()
        .route("/", get(get_schedules).post(create_schedule))
        .route("/:id", patch(update_schedule).delete(delete_schedule))
        .route_layer(axum::middleware::from_fn_with_state(
            state,
            is_authenticated,
        ))
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateSchedulePayload {
    pub date: DateTime<Utc>,
    pub start_time: String,
    pub end_time: String,
    pub title: String,
    pub subject: Option<String>,
    pub notes: Option<String>,
}

async fn get_schedules(
    State(state): State<Arc<AppState>>,
    Extension(auth_user): Extension<AuthUser>,
) -> Result<Json<Vec<Schedule>>, (StatusCode, String)> {
    let user_id = auth_user.0.id.unwrap();
    let coll = state.db.collection::<Schedule>("schedules");

    let mut cursor = coll
        .find(doc! { "userId": user_id })
        .await
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;

    let mut schedules = Vec::new();
    while let Some(item) = cursor.try_next().await.unwrap_or(None) {
        schedules.push(item);
    }

    Ok(Json(schedules))
}

async fn create_schedule(
    State(state): State<Arc<AppState>>,
    Extension(auth_user): Extension<AuthUser>,
    Json(payload): Json<CreateSchedulePayload>,
) -> Result<Json<Schedule>, (StatusCode, String)> {
    let user_id = auth_user.0.id.unwrap();
    let coll = state.db.collection::<Schedule>("schedules");

    let now = Utc::now();
    let new_item = Schedule {
        id: None,
        user_id,
        date: payload.date,
        start_time: payload.start_time,
        end_time: payload.end_time,
        title: payload.title,
        subject: payload.subject,
        notes: payload.notes,
        completed: false,
        created_at: now,
        updated_at: now,
    };

    let result = coll
        .insert_one(new_item.clone())
        .await
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;

    let mut inserted = new_item;
    inserted.id = result.inserted_id.as_object_id();

    Ok(Json(inserted))
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdateSchedulePayload {
    pub title: Option<String>,
    pub start_time: Option<String>,
    pub end_time: Option<String>,
    pub completed: Option<bool>,
}

async fn update_schedule(
    State(state): State<Arc<AppState>>,
    Extension(auth_user): Extension<AuthUser>,
    Path(id): Path<String>,
    Json(payload): Json<UpdateSchedulePayload>,
) -> Result<Json<Schedule>, (StatusCode, String)> {
    let user_id = auth_user.0.id.unwrap();
    let item_id =
        ObjectId::from_str(&id).map_err(|_| (StatusCode::BAD_REQUEST, "Invalid ID".into()))?;

    let coll = state.db.collection::<Schedule>("schedules");
    let filter = doc! { "_id": item_id, "userId": user_id };

    let mut update_doc = mongodb::bson::Document::new();
    if let Some(t) = payload.title {
        update_doc.insert("title", t);
    }
    if let Some(st) = payload.start_time {
        update_doc.insert("startTime", st);
    }
    if let Some(et) = payload.end_time {
        update_doc.insert("endTime", et);
    }
    if let Some(c) = payload.completed {
        update_doc.insert("completed", c);
    }

    update_doc.insert("updatedAt", Utc::now().to_rfc3339());

    if update_doc.is_empty() {
        return Err((StatusCode::BAD_REQUEST, "No update fields provided".into()));
    }

    coll.update_one(filter.clone(), doc! { "$set": update_doc })
        .await
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;

    let updated = coll
        .find_one(filter)
        .await
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?
        .ok_or((StatusCode::NOT_FOUND, "Not found".into()))?;

    Ok(Json(updated))
}

async fn delete_schedule(
    State(state): State<Arc<AppState>>,
    Extension(auth_user): Extension<AuthUser>,
    Path(id): Path<String>,
) -> Result<Json<serde_json::Value>, (StatusCode, String)> {
    let user_id = auth_user.0.id.unwrap();
    let item_id =
        ObjectId::from_str(&id).map_err(|_| (StatusCode::BAD_REQUEST, "Invalid ID".into()))?;

    let coll = state.db.collection::<Schedule>("schedules");
    coll.delete_one(doc! { "_id": item_id, "userId": user_id })
        .await
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;

    Ok(Json(serde_json::json!({ "success": true })))
}
